### PR TITLE
refactor: reuse protobuf Vertex definition

### LINKED ISSUE #151: Use `Vertex` from protobuf everywhere
Currently in the HashGraph we have a `Vertex` interface, should be replaced with the protobuf definition

### PR DESCRIPTION
**Pull request by cwkang1998**
Re-use exisiting protobuf Vertex definition for `packages/object`, removing the need to maintain a separate TS typedef. This is done by re-exporting the protobuf-gen typedef with the same name.

Since protobuf makes its message field optional by default since proto3, to address the undefined warning, a few modification is done:

1. `addToFrontier` now returns a Vertex with a defined `operation` field.
2. `merge` now ignores vertex with an undefined operation field. Alternatively we can make `merge` throw or log an error when vertex with undefined operation is received.

Closes #151.

### CODE REVIEW

**Comment by cwkang1998** [comment]
Might need to add some tests for the changes I make in `merge`, as its a logical change.

File: packages/object/src/hashgraph/index.ts [hunk 1]
```diff
@@ -86,10 +79,14 @@ export class HashGraph {
 		this.forwardEdges.set(HashGraph.rootHash, []);
 	}
 
-	addToFrontier(operation: Operation): Vertex {
+	/**
+	 * @param operation - The operation to be added
+	 * @returns The vertex that was added, with the operation being defined rather than undefined.
+	 */
+	addToFrontier(operation: Operation): Omit<Vertex, "operation"> & { operation: Operation } {
```
Comment (d-roak) [inline packages/object/src/hashgraph/index.ts:86]: Can safely return `Vertex`

File: packages/object/src/index.ts [hunk 2]
```diff
@@ -103,6 +104,11 @@ export class TopologyObject implements ITopologyObject {
 
 	merge(vertices: Vertex[]) {
 		for (const vertex of vertices) {
+			// Vertex operation is expected to be defined, if its not defined we ignore it and continue.
```
Comment (d-roak) [inline packages/object/src/index.ts:107]:
```suggestion
			// Check to avoid manually crafted `undefined` operations
```

File: packages/object/src/index.ts [hunk 3]
```diff
@@ -112,6 +118,10 @@ export class TopologyObject implements ITopologyObject {
 
 		const operations = this.hashGraph.linearizeOperations();
 		this.vertices = this.hashGraph.getAllVertices().map((vertex) => {
+			// Again, we are filtering out vertices with undefined operations.
```
Comment (d-roak) [inline packages/object/src/index.ts:121]:
```suggestion
```

File: packages/object/src/index.ts [hunk 4]
```diff
@@ -87,6 +87,7 @@ export class TopologyObject implements ITopologyObject {
 
 	// biome-ignore lint: value can't be unknown because of protobuf
 	callFn(fn: string, args: any) {
+		// Since we are creating a vertex with known values, the operation should not be undefined.
```
Comment (d-roak) [inline packages/object/src/index.ts:90]:
```suggestion
```

File: packages/object/src/hashgraph/index.ts [hunk 5]
```diff
@@ -86,10 +79,14 @@ export class HashGraph {
...
 		const deps = this.getFrontier();
 		const hash = computeHash(this.nodeId, operation, deps);
-		const vertex: Vertex = {
+		const vertex = {
 			hash,
 			nodeId: this.nodeId,
 			operation,
```
Comment (d-roak) [inline packages/object/src/hashgraph/index.ts:92]:
```suggestion
			operation ?? { type: OperationType.NOP },
```

**Review by d-roak — CHANGES_REQUESTED** [review]

File: packages/object/src/hashgraph/index.ts (hunk 1 above)
Comment (d-roak) [inline packages/object/src/hashgraph/index.ts:86]:
```suggestion
	addToFrontier(operation: Operation): Vertex {
```

File: packages/object/src/index.ts [hunk 6]
```diff
@@ -91,10 +91,7 @@ export class TopologyObject implements ITopologyObject {
 		const serializedVertex = ObjectPb.Vertex.create({
 			hash: vertex.hash,
 			nodeId: vertex.nodeId,
-			operation: {
-				type: vertex.operation.type,
-				value: vertex.operation.value,
-			},
+			operation: vertex.operation ?? { type: OperationType.NOP },
```
Comment (d-roak) [inline packages/object/src/index.ts:94]: no need to check type here, we can ensure it won't be null. If there is some edge case where it is, it'll be dropped by the peers and cease to exist

File: packages/blueprints/src/AddWinsSet/index.ts [hunk 7]
```diff
@@ -44,9 +44,12 @@ export class AddWinsSet<T> implements CRO {
 
 	// in this case is an array of length 2 and there are only two possible operations
 	resolveConflicts(vertices: Vertex[]): ResolveConflictsType {
+		// Both must have operations, if not return no-op
 		if (
-			vertices[0].operation.type !== vertices[1].operation.type &&
-			vertices[0].operation.value === vertices[1].operation.value
+			vertices[0].operation &&
+			vertices[1].operation &&
```
Comment (d-roak) [inline packages/blueprints/src/AddWinsSet/index.ts:50]: why is this needed? you changed correctly the other conditions e.g. `vertices[0].operation?....`

File: packages/object/src/index.ts [hunk 8]
```diff
@@ -111,17 +113,17 @@ export class TopologyObject implements ITopologyObject {
...
-					type: vertex.operation.type,
-					value: vertex.operation.value,
-				},
-				dependencies: vertex.dependencies,
-			};
-		});
+		this.vertices = this.hashGraph
+			.getAllVertices()
+			.filter((vertex) => vertex.operation !== undefined)
+			.map((vertex) => {
```
Comment (d-roak) [inline packages/object/src/index.ts:119]: can you test if we still need this mapping? Since we are using the same type everywhere now, it shouldn't be needed

File: packages/object/src/index.ts [hunk 9]
```diff
@@ -1,8 +1,8 @@
 import * as crypto from "node:crypto";
 import {
-	type ActionType,
 	HashGraph,
 	type Operation,
+	OperationType,
```
Comment (d-roak) [inline packages/object/src/index.ts:5]: import not needed, because of other comment

File: packages/object/tests/hashgraph.test.ts
Comment (d-roak) [inline packages/object/tests/hashgraph.test.ts:1]: love it !!

**Review by d-roak — CHANGES_REQUESTED** [review]

File: packages/object/src/index.ts (hunk 8 above)
Comment (cwkang1998) [reply packages/object/src/index.ts:119]: Yup, we don't need this anymore. Tests seems to pass.

File: packages/blueprints/src/AddWinsSet/index.ts (hunk 7 above)
Comment (d-roak) [reply packages/blueprints/src/AddWinsSet/index.ts:50]: in theory, it's impossible for that function to receive an undefined operation there (unless u mess with the hg manually). Although i really like your approach and it makes the code more reliable, we can go with it. Although it'll troublesome in the future, to detect that the linearizeOps is returning undefined/null ops

**Review by d-roak — APPROVED** [review]
lgtm
